teps/0057-windows-support.md

```diff
@@ -0,0 +1,154 @@
+
+### Goals
+- Allow Tekton TaskRuns to selectively run on Windows nodes
+- Allow a TaskRun to use nodeSelector to request this feature
+- Core Tekton components (Operator, CLI and Dashboard) must be able to recognise and communicate with Windows nodes in the same way they do Linux nodes
+- Tasks on the Catalog and Hub should be able to indicate the requirement for a Windows node to run it
+- All current (Linux-only) workflows should remain unchanged
+
+### Non-Goals
+- The ability to run core Tekton components on Windows nodes is not a goal of this TEP
+    - It is expected that users will have a mix of Linux and Windows nodes and that the core Tekton controller components will run on Linux nodes
+- Tekton will not be responsible for determining whether a Pod for a TaskRun should be created on a Windows node. This will be an opt-in feature determined via a nodeSelector.
+
+
+<!-- ### Use Cases (optional) -->
+
+<!--
+Describe the concrete improvement specific groups of users will see if the
+Motivations in this doc result in a fix or feature.
+
+Consider both the user's role (are they a Task author? Catalog Task user?
+Cluster Admin? etc...) and experience (what workflows or actions are enhanced
+if this problem is solved?).
+-->
+
+## Requirements
+Tekton should be able to:
+- Recognise and communicate with Windows nodes in the cluster
+- Allow TaskRuns to specify the need to run on a Windows node
+- Create Pods for TaskRuns on Windows nodes
+- Provide full support for all Tekton features to Windows based TaskRuns
+- Manage Pipelines containing a mix of Windows and Linux based Tasks
+
+Workflows that do not require Windows support should not be changed in any way.
+
+Tekton Hub and Catalog should allow a Task to indicate that it requires a Windows node to run, if this is absent, the default behaviour will be to assume Linux so that no change to existing Linux workflows is required.
+
+## Proposal
+Extend Tekton so TaskRuns can request, and be scheduled, to run on Windows nodes. This will require (at least) reworking the following Tekton components:
+- Entrypoint binary
+- All internal support images used for resource management
+- File paths used for step ordering, either via polling or watching files
+- Script mode
+
+### Notes/Caveats (optional)
+In order to have Windows nodes run TaskRuns there are some elements of Tekton that will need reworking to be compatible with that OS. Some of these have been identified in the [issue](https://github.com/tektoncd/pipeline/issues/1826) but it is almost certain that other elements will need work in order to work. 
+
+### Risks and Mitigations
+- The burden of supporting the Windows platform may be quite high
+- Windows containers may introduce some security vulnerabilities
+
+### User Experience (optional)
+- User specifies via nodeSelector if they want to run a task in a Windows container
```

Comment:
i dont know much about nodeSelectors and im wondering maybe you could provide an example - im wondering if the "nodeSelector" for windows nodes is always going to look the same, or if it could vary (i.e. will some clusters refer to their windows nodes as "win" and some refer to them as "windows", etc.) - because this user experience would suggest that the tekton pipelines controller needs to be able to look at the node selector and understand when it refers to a windows node (so it uses the appropriate images)

also afaik the nodeselector would likely be supplied at runtime (via a [pod template](https://github.com/tektoncd/pipeline/blob/1f5980f8c8a05b106687cfa3e5b3193c213cb66e/docs/podtemplates.md) in a TaskRun/Pipeline) - it seems like Tasks should have something in the Task spec that indicates if they are meant to be run on windows nodes or not

(let me know if this isnt making any sense - im not sure im explaining it well - im trying to refer to the difference between [runtime and authoring time](https://github.com/tektoncd/community/blob/main/design-principles.md#reusability) - this feels like an authoring time thing - which leads me to the conclusion that the nodeselector is maybe not the way the controller detects that it needs to use windows compatible images)

Reply:
I can't respond to the comment above, so I'll answer that here first. The big one is that it is possible to break-out of containers when using Windows (one example is documented [here](https://unit42.paloaltonetworks.com/windows-server-containers-vulnerabilities/). This is something that requires a fix from Microsoft, there's nothing that can be done at our end, bit obviously this is something we need to be conscious of moving forward.

Reply:
In terms of nodeSelectors - as I understand it kubernetes provides a fairly well defined way to require that a container be scheduled on a node that meets certain requirements. These can include OS requirements as described [here](https://kubernetes.io/docs/setup/production-environment/windows/user-guide-windows-containers/). I'd agree that we'd probably supply the nodeSelector at runtime, but we would definitely want something in the Task (i.e. authoring time)  that indicates this need. So maybe we do need something other than a nodeSelector here, but I'd imagine we'd want it to be defined along similar lines...

Another issue we need to face is the lack of flexibility when using windows containers on windows nodes - unless things have changed very recently then Windows _requires_ that the windows version of the node and container be the same.

Reply:
re. the vulnerabilites: thanks for the example! might be worth including that link in the content of the TEP as well

re. nodeSelectors:

Thanks for the explanation, I think that makes sense to me. My conclusion is that we ultimately will need to have something in the Task that indicates this need, and then at runtime folks will need to include the appropriate nodeselector for their cluster as well but (correct me if I'm wrong!!) Tekton probably can't be too opinionated about what the node selector looks like (since folks could be using any label they want to identify their windows nodes)? Or maybe there are already known conventions around identifying windows nodes we could embrace